src/node/p2p_message_handlers/mod.rs

```
@@ -64,16 +64,41 @@ pub async fn handle_request<C: 'static>(
             handle_share_headers(share_headers, chain_handle, time_provider).await
         }
         Message::ShareBlock(share_block) => {
-            if let Err(e) = handle_share_block(share_block, chain_handle, time_provider).await {
+            let peer_clone = peer.clone();
+            
+            // Define a simple closure that doesn't attempt to send through the channel
+            // This avoids the need to send the generic C type across threads
+            let disconnect_fn = Box::new(move |peer_to_disconnect: libp2p::PeerId| {
```

I am not too sure about this function, all it does is log a future disconnect. Do we really need to have a clouse and also change the signature of handle_share_block to log the event? Instead we should lean on error types. We should raise an error from validation, which when matched enqueues a disconnect into a disconnect task.

Something like the following will lead to more decoupled code and make it easier to maintain in the future.

1. Have a task waiting on a channel to disconnect peers.
2. On validation failure raise an error. [Add this error type].
3. When handle_share_block result is the specific error above, then enqueue the peer to the channel the disconnect task is listening to.
4. Drop connections and do any clean up in that task.